src/main/java/net/fabricmc/loader/api/VersionRange.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.api;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class VersionRange {
+	public enum Type {
+		INVALID,
+		ANY,
+		EQUALS,
+		GREATER_THAN,
+		LESSER_THAN,
+		GREATER_THAN_OR_EQUAL,
+		LESSER_THAN_OR_EQUAL,
+		SAME_MAJOR,
+		SAME_MAJOR_AND_MINOR;
+
+		public String represent(String version) {
+			switch (this) {
+			case INVALID:
+				return "unknown version";
+			case ANY:
+				return "any version";
+			case EQUALS:
+				return "version " + version;
+			case GREATER_THAN:
+				return "any version after " + version;
+			case LESSER_THAN:
+				return "any version before " + version;
+			case GREATER_THAN_OR_EQUAL:
+				return "version " + version + " or later";
+			case LESSER_THAN_OR_EQUAL:
+				return "version " + version + " or earlier";
+			case SAME_MAJOR:
+			case SAME_MAJOR_AND_MINOR:
+				String[] parts = version.split("\\.");
+				int start = 1;
+
+				if (this == Type.SAME_MAJOR_AND_MINOR) {
+					start = 2;
+				}
+
+				for (int i = start; i < parts.length; i++) {
+					parts[i] = "x";
+				}
+
+				return "version " + String.join(".", parts);
+			default:
+				return "unhandled version range type " + this;
```

Comment:
I'd make this throw an exception, or move the branches into the enum values themselves.

Reply:
https://github.com/FabricMC/fabric-loader/pull/277/commits/542fa57664e489bfb89fa6608e0317374f9dfbc1: it's a `UnaryOperator<String>` now i guess